### PR TITLE
NEW Wait for promises before submitting forms

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-elemental/issues/1215

### CODE REVIEW

File: client/src/legacy/LeftAndMain.js
Comment (emteknetnz) [inline client/src/legacy/LeftAndMain.js:473]: The diff in this PR looks confusing, though basically what I've done is nest the old logic instead the `Promise.all(...).then(...)` block

File: client/src/legacy/LeftAndMain.js
Comment (emteknetnz) [inline client/src/legacy/LeftAndMain.js:564]: This is a new line added

File: client/src/legacy/LeftAndMain.js
Comment (emteknetnz) [inline client/src/legacy/LeftAndMain.js:477]: This block is newly added - if any of the promises resolve(false) then the form will not submit

**Comment by emteknetnz** [comment]
Broken --prefer-lowest build is unrelated and will be fixed in https://github.com/silverstripe/gha-ci/issues/132

File: client/src/legacy/LeftAndMain.js
Comment (ChloeSartorelli) [reply client/src/legacy/LeftAndMain.js:473]: Diff is much more sensible with whitespace ignored 👌

File: client/src/legacy/LeftAndMain.js
Comment (ChloeSartorelli) [inline client/src/legacy/LeftAndMain.js]:
```suggestion
          if (result === false) {
```
I'd prefer this since that's [what you've said](https://github.com/silverstripe/silverstripe-admin/pull/1792#discussion_r1653845328) is the condition for this. Means we don't have to care about weird edge cases of other falsy values getting in here either with our own changes down the line or in peoples' customisations.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

**Review by GuySartorelli — APPROVED** [review]
LGTM, will merge when elemental PR is approved

File: client/src/components/HtmlEditorField/HtmlEditorField.js
Comment (emteknetnz) [inline client/src/components/HtmlEditorField/HtmlEditorField.js:57]: This change was made to resolve the "Blocks get double saved - which turns them back to "modified" issue noted [here](https://github.com/silverstripe/silverstripe-elemental/pull/1214#issuecomment-2218753147). The reason this was happening is that form-schema value of a blank HTMLText field (or Varchar field for that matter), starts of as `null`.

However as part of the HTMLEditorField initialising it has a `setcontent` call made to it, which in turns tells the redux that "this form has changes" and therefore `formDirty` is always set for the element, which is obviously wrong. `formDirty` is used in the new logic that determines if the element form should be inline submitted when clicking a page/parent save button.

I have regression tested this change in sink using this [PR](https://github.com/silverstripe/silverstripe-admin/pull/1796) that only contains this change.
That [run of sink](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/9983689111) has the same failures as a [vanilla sink run](https://github.com/silverstripe/recipe-kitchen-sink/actions/runs/9984678122)

**Review by GuySartorelli — APPROVED** [review]
LGTM, will merge when elemental PR is approved

File: client/src/boot/applyTransforms.js
Comment (emteknetnz) [inline client/src/boot/applyTransforms.js:73]: This is to resolve https://github.com/silverstripe/silverstripe-elemental/pull/1214#issuecomment-2237727252

I can't think of a clean way to abstract that isn't basically window.ss.excludeClientSideValidation = ['ElementForm_']

I think hardcoding may actually be better here as there's a nice big inline comment explaining what's going on.

Comment (ChloeSartorelli) [reply client/src/boot/applyTransforms.js:73]: Hardcoding isn't ideal but it gets the job done. Would be better if we did something like `if (field.parents('.skipClientSideFormValidation').length > 0)` (if any parent of the field has that css class, then we can skip this client-side validation) and add a `skipClientSideFormValidation` css class to the elemental form.

I'll let you decide if you want to do that - if not I'm okay with merging as is.

Comment (emteknetnz) [reply client/src/boot/applyTransforms.js:73]: I think merge as.

I've had a quick look at doing this "properly", would involved adding new API to formschema rather than just a class on a form.  Also because it deals with validation I think we'd possibly want to add it to the CompositeValidator, rather than directly on the Form e.g. Form->setClientSideValidation(false);  I'm not sure where it would be best.

Regardless it would be a proper standalone PR.  We could easily add this later should the need arise and remove the hack added here, so I think the hack here is OK

Comment (ChloeSartorelli) [reply client/src/boot/applyTransforms.js:73]: Ahhh it's react nonsense. That explains why this doesn't happen for validation of fields directly on the page. I thought this was jquery nonsense but honestly it being react both makes sense and is hilarious.
No worries.

**Review by GuySartorelli — APPROVED** [review]
LGTM, will merge when elemental PR is approved

**Comment by GuySartorelli** [comment]
CI failure will be resolved as part of https://github.com/silverstripe/.github/issues/285

**Comment by GuySartorelli** [comment]
Will rerun CI after https://github.com/silverstripe/silverstripe-admin/pull/1792 is merged

**Comment by GuySartorelli** [comment]
Can you please rebase this? Looks like it needs a rebase to pick up the missing `@job` tags in behat.

**Comment by emteknetnz** [comment]
Done

**Review by GuySartorelli — APPROVED** [review]
LGTM

**Pull request merged**
